scripts/cube_fx.lua

```diff
@@ -304,26 +308,122 @@ local function cube_victory(size, results)
   end
 end
 
+-- Map to determine if some machine is using the cube
+-- in an "efficient" manner. 
+local cube_machine_condition_handlers = {
+  ["status"] = function(entity, cube_type)
+    return entity.status == defines.entity_status.working
+  end,
+  ['burner'] = function(entity, cube_type)
+    if entity.status ~= defines.entity_status.working then return false end
+    local burner = entity.burner
+    if not burner then return false end
+    -- Now we know this entity is currently burning
+    if burner.currently_burning and burner.currently_burning.name == cube_type then return true end
+    return false -- This burner is burning something else
+  end,
+  ["handcrafting"] = function(character, cube_type)
+    -- We could search through the crafting queue, but this function
+    -- needs to be quick because it's called every 6 ticks. Lets
+    -- rather assume that if the player doesn't have it in their
+    -- inventory or cursor then the player is crafting with it.
+    -- That doesn't handle it being in queues though, but it's
+    -- better than nothing.
+    local player = character.player
+    if not player then return end
+    local cursor_stack = player.cursor_stack
+    if cursor_stack.valid and cursor_stack.valid_for_read then
+      if cursor_stack.name == cube_type then return false end
+    end
+    local inventory = player.get_main_inventory()
+    if inventory.find_item_stack(cube_type) then return false end
+    return true
+  end,
+}
+
+-- Map to determine in which machine types the cube can be
+-- seen as being "utilization". The value will show what 
+-- requirement this machine has to be "productive" with
+-- the cube. 
+local cube_working_machine_types = {
+  ["character"]           = cube_machine_condition_handlers["handcrafting"],
+
+  -- For the following entity-types we only ever have to check the entity status.
+  -- This is because in 99% of cases if the cube can be placed in the machine
+  -- then it has to be an ingredient or fuel. And thus, if the cube is in a machine
+  -- and it's not currently "working" then it would mean the cube is idling due to
+  -- `no_fuel`, `output_full`, `no_ingredients`, etc.
+  -- One edge case might be phantom cube placed in the output slot of a boiler
+  -- while the boiler is burning, but the only possible fuel source then
+  ["assembling-machine"]  = cube_machine_condition_handlers["status"],
+  ["furnace"]             = cube_machine_condition_handlers["status"],
+  ["generator"]           = cube_machine_condition_handlers["status"],
+  ["burner-generator"]    = cube_machine_condition_handlers["status"],
+  ["rocket-silo"]         = cube_machine_condition_handlers["status"],
```

Comment:
I'm not sure what the timing is like using the teleporter, but if there's a launching animation we could set it as `used` while the animation is ongoing? Can't play with it though cause it currently crashes :)

Reply:
Just played with it a little and IMO this should count as `idle`, because the cube is only being transported, not used The same reason IMO we're ignoring inserters, cars, trains, etc. So for now I'm removing the `rocket-silo` entry, so that it's always counted as idle.